README.md
# Pažymių vidurkio skaičiuoklė 🧮

Programa yra skirta apskaičiuoti pažymių vidurkį ir/arba medianą.

## Turinys

- [🥤 Veikimo principas](#-veikimo-principas)
  - [📚 Generate new records](#-generate-new-records)
  - [🌪️ Filter records](#%EF%B8%8F-filter-records)
  - [📂 Read grades from a file](#-read-grades-from-a-file)
  - [✍️ Enter grades manually](#%EF%B8%8F-enter-grades-manually)
- [📈 Spartos analizė](#-spartos-analizė)
  - [🛠️ Duomenų apdorojimas](#%EF%B8%8F-duomenų-apdorojimas)
  - [⚖️ Duomenų rūšiavimas](#%EF%B8%8F-duomenų-rūšiavimas)
    - [Strategija 1](#strategija-1)
    - [Strategija 2](#strategija-2)
    - [Strategija 3](#strategija-3)
- [🚀 Programos diegimas ir paleidimas](#-programos-diegimas-ir-paleidimas)

---

## 🥤 Veikimo principas

> Paleidus programą, prašome sekti komandinėje eilutėje nurodytas instrukcijas.

```bash
1. Generate new records
2. Filter records
3. Read grades from a file
4. Enter grades manually
-> Select: (1-4):
```

1. [Generate new records](#-generate-new-records)
2. [Filter records](#%EF%B8%8F-filter-records)
3. [Read grades from a file](#-read-grades-from-a-file)
4. [Enter grades manually](#%EF%B8%8F-enter-grades-manually)

---

### 📚 Generate new records

Šis metodas skirtas naujų failų generavimui.

```bash
-> Select: (1-4): 1
-> How many records: (1-10000000): 1000
```

Po įvedimo `./data/` direktorijoje yra sukuriamas naujas failas `Random XXX.txt`, kur `XXX` yra užklausos dydis.
Jame sugeneruojami studentų įrašai su 10 atsitiktinių namų darbų įvertinimų bei egzamino pažymiu.

```bash
Vardas     Pavarde     ND1  ND2  ND3 ... ND10  Egz.
Vardas1    Pavarde1    2    8    6       9     3
Vardas2    Pavarde2    1    2    3       7     10
...
VardasXXX  PavardeXXX  3    4    8       6     8
```

Žemiau pateikti testavimo atvejai. Laiko matavimas pateiktas sekundėmis.

| Description        | 1000    | 10.000  | 100.000 | 1.000.000 | 10.000.000 |
| :----------------- | :------ | :------ | :------ | :-------- | :--------- |
| Buffering students | 0.00344 | 0.02834 | 0.23464 | 2.27576   | 23.217     |
| Saving buffer      | 0.00049 | 0.00163 | 0.01616 | 0.13867   | 1.56499    |
| **Total**          | 0.00398 | 0.03003 | 0.25170 | 2.42024   | 24.8665    |

---

### 🌪️ Filter records

Šis metodas skirtas atskirti gerai besimokančius studentus nuo prastai besimokančių. Jei studento **vidurkis** yra didesnis arba lygus nei 5, tai jis priskiriamas prie gerai besimokančių, kitu atveju - prie besimokančių prastai.

```bash
-> Select: (1-4): 2
Folder "./data/" contains multiple files. Please select one:
1: Studentai 10.000.txt
2: Studentai 100.000.txt
3: Studentai 1000.000.txt
4: Studentai 10.000.000.txt
-> Please select a file: (1-4): 1
```

Gerai besimokinantys studentai išsaugomis faile `XXX winners.txt`, o prastai - `XXX losers.txt`, kur `XXX` yra šakninis pasirinkto failo pavadinimas.

Žemiau pateikti testavimo atvejai. Laiko matavimas pateiktas sekundėmis.

| Description                    | 10.000  | 100.000 | 1.000.000 | 10.000.000 |
| :----------------------------- | :------ | :------ | :-------- | :--------- |
| **Reading source file:**       |         |         |           |            |
| Buffering file                 | 0.01779 | 0.18051 | 0.80284   | 7.6818     |
| Processing buffer              | 0.06277 | 0.71893 | 3.57932   | 36.7005    |
| Processing students            | 0.00024 | 0.00234 | 0.03426   | 0.27276    |
| **Filtering:**                 |         |         |           |            |
| Sorting students (descending)  | 0.00076 | 0.01204 | 0.09973   | 1.02805    |
| Searching for the first loser  | 2.9e-05 | 0.00164 | 0.00378   | 0.04699    |
| Copying losers to a new vector | 0.00069 | 0.01164 | 0.07389   | 0.92133    |
| Resizing original vector       | 0.00036 | 0.01485 | 0.09852   | 1.26067    |
| **Writing losers to file:**    |         |         |           |            |
| Buffering students             | 0.01385 | 0.15100 | 0.59066   | 5.91808    |
| Writing buffer to file         | 0.00105 | 0.01275 | 0.04718   | 0.49510    |
| **Writing winners to file:**   |         |         |           |            |
| Buffering students             | 0.01886 | 0.21706 | 0.86465   | 8.9698     |
| Writing buffer to file         | 0.00181 | 0.01645 | 0.06368   | 0.68474    |
| **Total**                      | 0.11841 | 1.34499 | 6.27838   | 64.1041    |

---

### 📂 Read grades from a file

Šis metodas skirtas **vidurkio** arba **mediano** (arba jų abiejų) skaičiavimui iš pasirinkto failo.

Programa ieško `*.txt` failų `./data/` direktorijoje. Aptikus kelis failus, programa siūlo pasirinkti, kurį failą nuskaityti:

```bash
-> Select: (1-4): 3
Folder "./data/" contains multiple files. Please select one:
1: Studentai 10.000.txt
2: Studentai 100.000.txt
3: Studentai 1000.000.txt
4: Studentai 10.000.000.txt
-> Please select a file: (1-4): 1
Reading data from "./data/Studentai 10.000.txt"
-> Choose what to calculate: (1)Mean, (2)Median, (3)Both: (1-3): 3
```

Po sėkmindo skaičiavimo duomenis galima automatiškai išsaugoti į failą, arba pateikti juos į konsolę:

```bash
-> Choose what to calculate: (1)Mean, (2)Median, (3)Both: (1-3): 3
-> Print to (y)file or (n)console: (y/n): y
```

Pasirinkus išsaugojimą į failą, rezultatas atrodo taip:

```bash
Vardas     Pavarde     Galutinis Vid.  Galutinis Med.
-----------------------------------------------------
Vardas1    Pavarde1    3.43            6.00
Vardas10   Pavarde10   7.23            4.00
...
VardasXXX  PavardeXXX  6.45            7.00
```

---

### ✍️ Enter grades manually

Metodas skirtas studento duomenims suvesti rankiniu būdų. Programa siūlo sekančius pasirinkimus:

- Įvesti vardą ir pavardę,
- Įvesti namų darbų įvertinimus bei egzamino balą.
- Po duomenų suvedimo yra apskaičiuojamas įvertinimų vidurkis arba medianas.

> Programa suteikia galimybę generuoti atsitiktinius namų darbų rezultatus bei egzamino balą, bei suvesti daugiau nei vieno studento duomenis.

Vienas iš rankiniu būdu suvestos informacijos programos eigos variantų gali būti toks:

```shell
-> Select: (1-4): 3
-> Please enter first name: Tomas
-> Please enter last name: Makaronas
-> Do you know the number of grades? (y/n): y
Enter number of grades: 5
-> Generate RANDOM grades (otherwise, enter grades MANUALLY)? (y/n): n
Enter grades: 7 8 9 9 10
Enter exam grade: 10
-> Add another student? (y/n): y
-> Please enter first name: Jurga
-> Please enter last name: Spurga
-> Do you know the number of grades? (y/n): n
-> Generate RANDOM grades (otherwise, enter grades MANUALLY)? (y/n): y
Generated 7 random grades: 8 5 3 8 2 8 5
Generated random exam grade: 4
-> Add another student? (y/n): y
-> Please enter first name: Studentas
-> Please enter last name: Nabagelis
-> Do you know the number of grades? (y/n): n
-> Generate RANDOM grades (otherwise, enter grades MANUALLY)? (y/n): n
Enter grade [0] (type -1 to quit): 7
Enter grade [1] (type -1 to quit): 8
Enter grade [2] (type -1 to quit): 9
Enter grade [3] (type -1 to quit): -1
Enter exam grade: 8
-> Add another student? (y/n): n
-> Choose what to calculate: (1)Mean, (2)Median, (3)Both: (1-3): 3
-> Print to (y)file or (n)console: (y/n): n
```

Po sėkmingo duomenų suvedimo į terminalo langą išvedami rezultatai panašia forma:

```shell
Vardas      Pavarde         Galutinis Vid. Galutinis Med.
----------------------------------------------------------
Tomas       Makaronas       9.44           9.00
Studentas   Nabagelis       8.00           8.00
Jurga       Spurga          4.63           5.00
```

Galutinis vidurkis yra apskaičiuojamas pagal formulę `galutinis = 0.4 * vidurkis + 0.6 * egzaminas`.

---

## 📈 Spartos analizė

Programos spartos analizė (duomenų, nuskaitymas iš failo, apdorijimas ir rūšiavimas į kategorijas) atlikta duomenis saugant į skirtingus konteinerius: `std::deque`, `std::list` ir `std::vector`.

Analizė atlikta su MacBook Pro, 2.3 GHz Dual-Core Intel Core i5, 8 GB 2133 MHz LPDDR3.

---

### 🛠️ Duomenų apdorojimas

| Container   | 10.000  | 100.000 | 1.000.000 | 10.000.000 |
| :---------- | :------ | :------ | :-------- | :--------- |
| std::deque  | 0.06121 | 0.71367 | 3.53594   | 35.27019   |
| std::list   | 0.06223 | 0.71956 | 3.62151   | 35.82814   |
| std::vector | 0.06038 | 0.71421 | 3.52655   | 36.21650   |

---

### ⚖️ Duomenų rūšiavimas

#### Strategija 1

Konteinerio (`vector`, `list` ir `deque`) skaidymas į **du naujus to paties tipo konteinerius**: _losers_ ir _winners_. Šiuo būdu tas pats studentas yra dviejuose masyvuose: pagrindiniame ir viename iš suskaidytų.

| Container / Task              | 10.000  | 100.000 | 1.000.000 | 10.000.000 |
| :---------------------------- | :------ | :------ | :-------- | :--------- |
| std::deque: Sorting students  | 0.00087 | 0.01414 | 0.12721   | 1.25897    |
| std::list: Sorting students   | 0.00172 | 0.02298 | 0.17643   | 1.77439    |
| std::vector: Sorting students | 0.00109 | 0.01884 | 0.21413   | 1.98630    |

---

#### Strategija 2

Konteinerio (`vector`, `list` ir `deque`) skaidymas panaudojant **tik vieną naują konteinerį**: _losers_. Tokiu būdu, jei studentas yra _loser'is_, jį perkeliame į naująjį _losers_ masyvą ir ištriname iš bendro masyvo. Po šio žingsnio pagrindiniame masyve lieka tik _winner'isi_.

| Container / Task                 | 10.000  | 100.000 | 1.000.000  | 10.000.000 |
| :------------------------------- | :------ | :------ | :--------- | :--------- |
| std::deque: Sorting and erasing  | 0.05147 | 5.16963 | 617.121 💩 | ---        |
| std::list: Sorting and erasing   | 0.00095 | 0.01309 | 0.11558    | 1.07584    |
| std::vector: Sorting and erasing | 0.12405 | 14.4877 | 2095.87 💩 | ---        |

---

#### Strategija 3

Konteinerio (`vector`, `list` ir `deque`) skaidymas panaudojant **tik vieną naują konteinerį**: _losers_. Ši strategija yra imlesnė darbui, o jos veikimo principas yra sekantis:

1. Išrūšiuoti studentų masyvą pagal galutinį pažymį mažėjimo tvarka,
2. Rasti indeksą, nuo kurio studentų balai yra mažesni nei `5`,
3. Perkopijuoti elementus nuo rasto indekso iki masyvo pabaigos į naująjį masyvą,
4. Sumažinti pagrindinio masyvo dydį nuo `0` iki rasto indekso.

| Container / Task               | 10.000  | 100.000 | 1.000.000 | 10.000.000 |
| :----------------------------- | :------ | :------ | :-------- | :--------- |
| std::deque                     |
| Sorting students (descending)  | 0.00110 | 0.01302 | 0.10790   | 1.12506    |
| Searching for the first loser  | 0.00003 | 0.00045 | 0.00404   | 0.04759    |
| Copying losers to a new vector | 0.00052 | 0.01091 | 0.08988   | 0.91767    |
| Resizing original vector       | 0.00037 | 0.01114 | 0.10823   | 1.34660    |
| **Total**                      | 0.00202 | 0.03552 | 0.31005   | 3.43692    |
| std::list                      |
| Sorting students (descending)  | 0.00167 | 0.05243 | 0.74777   | 11.87566   |
| Searching for the first loser  | 0.00009 | 0.00419 | 0.06602   | 0.62471    |
| Copying losers to a new vector | 0.00073 | 0.01167 | 0.12417   | 1.31286    |
| Resizing original vector       | 0.00089 | 0.02170 | 0.22703   | 2.81853    |
| **Total**                      | 0.00338 | 0.08999 | 1.16499   | 16.63176   |
| std::vector                    |
| Sorting students (descending)  | 0.00087 | 0.01300 | 0.10440   | 1.05385    |
| Searching for the first loser  | 0.00002 | 0.00036 | 0.00354   | 0.04925    |
| Copying losers to a new vector | 0.00053 | 0.01058 | 0.07346   | 0.92085    |
| Resizing original vector       | 0.00036 | 0.01487 | 0.09988   | 1.26714    |
| **Total**                      | 0.00178 | 0.03881 | 0.28128   | 3.29109    |

---

## 🚀 Programos diegimas ir paleidimas

- Atsisiųskite programos versiją iš [Releases](https://github.com/rendertom/VU-OP-Task-2-Pazymiu-skaiciuokle/releases) aplanko ir ją išsiarchyvuokite,
- Komandinėje eilutėje įvykdykite sekančius veiksmus:

```shell
cd nuoroda_į_programos_aplanką
g++ -std=c++11 -O3 include/*.cpp 'main.cpp' -o 'main'
./main
```


changelog.md
# Changelog

- [v0.1](https://github.com/rendertom/VU-OP-Task-2-Pazymiu-skaiciuokle/releases/tag/v0.1) - Pradinė programos versija
- [v0.1.1](https://github.com/rendertom/VU-OP-Task-2-Pazymiu-skaiciuokle/releases/tag/v0.1.1)
  - pataisyta
    - Vidurkio ir Mediano skaičiavimas, kai nėra įvesta pažymių.
    - Nerodo pažymių įvedimo formos užklausos (random/manual), jei vartotojas neturi pažymių.
  - pašalinta
    - VSCode IDE nustatymai (.vscode > settings.json).
    - `std::cout` teksto spalvos nustatymai.
- [v0.2](https://github.com/rendertom/VU-OP-Task-2-Pazymiu-skaiciuokle/releases/tag/v0.2)
  - pataisyta
    - Nerodo atsitiktinės (random) pažymių įvedimo užklausos, kai pažymių skaičius yra nežinomas.
  - pašalinta
    - `C` tipo masyvo programos realizcija: pašalintas failas `main-array.cpp`, o failas `main-vector.cpp` pervadintas į `main.cpp`.
  - pridėta
    - Galimybė skaityti rezultatus iš išorinio failo.
    - Galimybė atspausdinti Vidurkį, Medianą, arba abu iš karto.
    - Spausdinamas sarašas išrikiuojamas abėcėlės tvarka pagal pavardes.
- [v0.3](https://github.com/rendertom/VU-OP-Task-2-Pazymiu-skaiciuokle/releases/tag/v0.3)
  - pakeista
    - Optimizuotas nuskaitomo failo apdorojimas, nuskaitant jo turinį į buffer'į.
    - Optimizuotas ciklo _for_ veikimas, nuskaitant masyvo dydį į kintamąjį.
    - Reorganizuota failo struktūra, suskaldant pagrindinį failą į smulkesnius failus su antraštėm _header_. Sukurti atskiri _header_ failai: Console.hpp, Definitions.hpp, File.hpp, Math.hpp, RND.hpp, Student.hpp, Table.hpp, Timer.hpp.
  - pridėta
    - Išimčių _exceptions_ valdymas.
    - Failų, esančių `/data/`direktorijoje pasirinkimas nuskaitymui naudojant komandinę eilutę.
- [v0.4](https://github.com/rendertom/VU-OP-Task-2-Pazymiu-skaiciuokle/releases/tag/v0.4)
  - pridėta
    - Galimybė generuoti naujus studentų sarašų failus.
    - Galimybė filtruoti studentus ir išsaugoti rezultatus į atskirus failus.
- [v0.5](https://github.com/rendertom/VU-OP-Task-2-Pazymiu-skaiciuokle/releases/tag/v0.5)
  - pridėta
    - Galimybė pasirinkti duomenų saugojimo konteinerį: `std::deque`, `std::list` arba `std::vector`.
- [v0.5.1](https://github.com/rendertom/VU-OP-Task-2-Pazymiu-skaiciuokle/releases/tag/v0.5.1)
  - pakeista
    - Konteinerių kūrimo principas.
- [v1.0](https://github.com/rendertom/VU-OP-Task-2-Pazymiu-skaiciuokle/releases/tag/v1.0)
  - pridėta
    - Galimybė pasirinkti filtravimo strategiją.


main-array.cpp
#include <algorithm>  //std::sort
#include <iomanip>    // std::setw
#include <iostream>
#include <random>  // std::rand, std::srand
#include <string>

#define GRADE_MIN 1
#define GRADE_MAX 10

using std::cin;
using std::cout;
using std::endl;
using std::fixed;
using std::left;
using std::numeric_limits;
using std::setprecision;
using std::setw;
using std::sort;
using std::streamsize;
using std::string;

struct Student {
  string firstName;
  string lastName;
  int numGrades;
  int *grades;
  int examGrade;
  double meanGrade;
  double finalGrade;
  double medianGrade;
};

void arrayCopy(int *targetArray, int *sourceArray, int size) {
  for (int i = 0; i < size; i++) {
    targetArray[i] = sourceArray[i];
  }
}

void arrayPush(int *&array, int &size, int value) {
  int *tempArray = new int[size + 1];
  arrayCopy(tempArray, array, size);
  tempArray[size] = value;

  delete[] array;
  array = tempArray;
  size++;
}

void clearLine() {
  cin.clear();
  cin.ignore(numeric_limits<streamsize>::max(), '\n');
}

bool confirm(const string &message, char yes = 'y', char no = 'n') {
  while (true) {
    cout << "-> " << message << " (" << yes << "/" << no << "): ";

    char response;
    cin >> response;
    clearLine();
    if (response == yes) {
      return true;
    } else if (response == no) {
      return false;
    } else {
      cout << "Unknown character. ";
    }
  }
}

double findMean(int *array, int length) {
  int sum = 0;
  for (int i = 0; i < length; i++) {
    sum += array[i];
  }

  return (double)sum / length;
}

double findMedian(int *array, int arrayLength) {
  sort(array, array + arrayLength);

  bool isOddNumber = arrayLength % 2 != 0;

  if (isOddNumber) {
    return (double)array[arrayLength / 2];
  }

  return (double)(array[(arrayLength - 1) / 2] + array[arrayLength / 2]) / 2.0;
}

void freeMemory(Student *students, int numStudents) {
  for (int i = 0; i < numStudents; i++) {
    delete[] students[i].grades;
  }
  delete[] students;
}

int getNumberOfGrades() {
  cout << "Enter number of grades: ";
  int numGrades;
  cin >> numGrades;
  clearLine();

  while (numGrades < 0) {
    cout << "Value cannot be negative. Please enter new value: ";
    cin >> numGrades;
    clearLine();
  }

  return numGrades;
}

int getRandomIntegerInRange(int min, int max) {
  static bool first = true;
  if (first) {
    srand(time(NULL));  //seeding for the first time only!
    first = false;
  }
  return min + rand() % ((max + 1) - min);
}

bool isValidGrade(int grade) {
  return grade >= GRADE_MIN && grade <= GRADE_MAX;
}

void printResult(Student *student, bool showMeanGrade = true) {
  cout << left
       << setw(10) << student->firstName
       << setw(15) << student->lastName
       << setw(12) << fixed << setprecision(2)
       << (showMeanGrade ? student->finalGrade : student->medianGrade)
       << endl;
}

void printResults(Student *students, int numStudents, bool showMeanGrade = true) {
  cout << left
       << setw(10) << "Vardas"
       << setw(16) << "Pavardė"
       << "Galutinis " << (showMeanGrade ? "Vid." : "Med.") << endl;
  cout << "-----------------------------------------------------------" << endl;

  for (int i = 0; i < numStudents; i++) {
    printResult(&students[i], showMeanGrade);
  }
}

void printStudent(Student *student) {
  cout << "First name: " << student->firstName << endl;
  cout << "Last name: " << student->lastName << endl;
  cout << "Number of grades: " << student->numGrades << endl;

  cout << "Grades: ";
  for (int i = 0; i < student->numGrades; i++) {
    cout << student->grades[i] << " ";
  }
  cout << endl;

  cout << "Exam grade: " << student->examGrade << endl;
  cout << "Mean grade: " << student->meanGrade << endl;
  cout << "Final grade: " << student->finalGrade << endl;
  cout << "Median grade: " << student->medianGrade << endl;
}

void processStudent(Student *student, bool shouldCalculateMean = true) {
  student->finalGrade = 0;
  student->meanGrade = 0;
  student->medianGrade = 0;
  if (shouldCalculateMean) {
    if (student->numGrades > 0) {
      student->meanGrade = findMean(student->grades, student->numGrades);
    }
    student->finalGrade = 0.4 * student->meanGrade + 0.6 * student->examGrade;
  } else {
    if (student->numGrades > 0) {
      student->medianGrade = findMedian(student->grades, student->numGrades);
    }
  }
}

void printRandomGrades(Student *student) {
  cout << "Generated " << student->numGrades << " random grades: ";
  for (int i = 0; i < student->numGrades; i++) {
    cout << student->grades[i] << " ";
  }
  cout << endl;
  cout << "Generated random exam grade: " << student->examGrade << endl;
}

void pushStudent(Student *&students, int &numStudents, Student student) {
  Student *tempArray = new Student[numStudents + 1];

  for (int i = 0; i < numStudents; i++) {
    tempArray[i] = students[i];
  }

  tempArray[numStudents] = student;

  delete[] students;
  students = tempArray;
  numStudents++;
}

int main() {
  int numStudents = 0;
  Student *students = new Student[numStudents];

  while (true) {
    Student student;

    cout << "Please enter first name: ";
    getline(cin, student.firstName);

    cout << "Please enter last name: ";
    getline(cin, student.lastName);

    bool numberOfGradesIsKnown = confirm("Do you know the number of grades?");
    student.numGrades = numberOfGradesIsKnown ? getNumberOfGrades() : 0;
    student.grades = new int[student.numGrades];

    bool shouldGenerateRandomGrades = false;
    if (student.numGrades > 0) {
      shouldGenerateRandomGrades = confirm("Generate RANDOM grades (otherwise, enter grades MANUALLY)?");
    }

    if (shouldGenerateRandomGrades) {
      if (numberOfGradesIsKnown) {
        for (int i = 0; i < student.numGrades; i++) {
          student.grades[i] = getRandomIntegerInRange(GRADE_MIN, GRADE_MAX);
        }
      } else {
        while (true) {
          int grade = getRandomIntegerInRange(0, GRADE_MAX);
          if (isValidGrade(grade)) {
            arrayPush(student.grades, student.numGrades, grade);
          } else {
            break;
          }
        }
      }
      student.examGrade = getRandomIntegerInRange(GRADE_MIN, GRADE_MAX);
      printRandomGrades(&student);
    } else {
      if (numberOfGradesIsKnown) {
        if (student.numGrades > 0) {
          cout << "Enter grades: ";
          int gradeIndex = 0;
          while (gradeIndex != student.numGrades) {
            int grade;
            cin >> grade;
            if (!isValidGrade(grade)) {
              cout << "Grade " << grade << " at index " << gradeIndex << " is out of range ("
                   << GRADE_MIN << "-" << GRADE_MAX << "). Fix it and continue entering." << endl;
              clearLine();
            } else {
              student.grades[gradeIndex] = grade;
              gradeIndex++;
            }
          }

          clearLine();
        }
      } else {
        while (true) {
          cout << "Enter grade [" << student.numGrades << "] (type -1 to quit): ";

          int grade;
          cin >> grade;
          clearLine();

          if (grade == -1) {
            break;
          } else {
            if (!isValidGrade(grade)) {
              cout << "Grade " << grade << " is out of range ("
                   << GRADE_MIN << "-" << GRADE_MAX << ")." << endl;
            } else {
              arrayPush(student.grades, student.numGrades, grade);
            }
          }
        }
      }

      cout << "Enter exam grade: ";
      cin >> student.examGrade;
      while (!isValidGrade(student.examGrade)) {
        cout << "Grade " << student.examGrade << " is out of range ("
             << GRADE_MIN << "-" << GRADE_MAX << "). Fix it and continue entering." << endl;
        clearLine();
        cin >> student.examGrade;
      }

      clearLine();
    }

    pushStudent(students, numStudents, student);
    if (!confirm("Add another student?")) {
      break;
    }
  }

  bool shouldCalculateMean = confirm("Calculate MEAN (otherwise, calculate MEDIAN)?");
  for (int i = 0; i < numStudents; i++) {
    processStudent(&students[i], shouldCalculateMean);
  }

  cout << endl;
  printResults(students, numStudents, shouldCalculateMean);
  cout << endl;

  freeMemory(students, numStudents);

  return 0;
}